Proposal of a source code inventory pattern (#322)
gruetter/innerSourceTest: patterns/1-initial/source-code-inventory.md

## Title

> Unified source code inventory

Reviewer: ```suggestion
Unified Source Code Inventory
```

Let's go for title-case here, as that's how the other patterns do it.

Reviewer: This seems non-controversial. Merging.

## Patlet

> In a large organization with different legal entities is often hard to get full visibility into all software assets, in particular all source code. This situation reduces the opportunities to increase business value and keep liability costs, such as software maintenance, under control across the organization as a whole. An organization level source code inventory addresses these issues while exploiting opportunities to identify and support valuable InnerSource assets.

Reviewer: ```suggestion
In a large organization with different legal entities is often hard to get full visibility into all software assets, in particular all source code. This situation reduces the opportunities to increase business value and keep liability costs, such as software maintenance, under control across the organization as a whole. An organization-level source code inventory addresses these issues while exploiting opportunities to identify and support valuable InnerSource assets.
```

Reviewer: I believe that this is how it should be spelled. If you agree, then there are a couple of more similar instances in the rest of the doc to fix accordingly.

* You do not have a clear policy on default sharing level when creating a new project (e.g., Open Source, InnerSource or Closed Source).
* You cannot scan significant parts of the Organization's source code looking for duplication, similarity or code smells.
* You do not know the existing ratios of Open Source, innerSource and Closed Source and their trend.
* You cannot measure the diversity of contributions and resulting value for a given project.
* You cannot identify and optimize tech stack diversity.
* You cannot identify technical debt and determine the priorities for retirement (e.g., dead APIs/source).

## Forces

* Fragmentation of source code hosting systems in the organization.
* Ad-hoc source code strategies scattered across the different Legal Entities in the organization.
* Continuously changing map of the relationships between: projects, repositories, products, tech stacks, domains, solutions, platforms, services, components, sub-systems, people, authors, teams, external repositories.
* Diverse software culture of teams across the organization (e.g., more open to collaboration or more siloed).

## Solutions

> ### Set up an Organization level live source code inventory

Reviewer: What do you mean by "live" here?

Author: I mean a live dashboard. I've updated to make it consistent with the snapshots

* Key meta-data about each repository:
  * Legal Entity
  * URL
  * Version control system (e.g., GIT or SVN).
  * Hosting vendor (e.g., GitHub, Gitlab or BitBucket) and hosting type (e.g., on-prem, private cloud or public cloud).
  * Sharing level (e.g., Open Source, InnerSource, Closed Source).
* Visualization in place to list all assets with options to filter based on meta-data
* Enable access to automated source code static analysis tools (e.g., identify duplicated or similar code, flag code smells, benchmark test coverage).

Mockup dashboard             |  Mockup questionnaire
:-------------------------:|:-------------------------:
![Mockup dashboard](../../assets/img/source-code-inventory-mockup-dashboard.PNG)  |  ![Mockup questionnaire](../../assets/img/source-code-inventory-mockup-questionnaire.PNG)

### Define a Source Code Strategy Assessment Framework

* Help to define a source code strategy including an explicit definition of policies at Organization and Legal Entity levels for:
  * Inventory
  * Sharing
  * Consumption
  * Contribution
  * Ecosystem

> * Example of such a framework: [https://github.com/trieshard/source-strategy-assessment/blob/master/framework.md](https://github.com/trieshard/source-strategy-assessment/blob/master/framework.md)

Reviewer: ```suggestion
* Example of such a framework: [github.com/trieshard/source-strategy-assessment/blob/master/framework.md](https://github.com/trieshard/source-strategy-assessment/blob/master/framework.md)
```

* We can identify technical debt and determine the priorities for retirement (e.g., dead APIs/source).

### For InnerSource governance

* We can scan significant parts of our Organization's source code looking for opportunities of reuse, duplication, similarity or code smells.
* We know the ratios of Open Source, InnerSource and Closed Source within the Organization so we can steer as needed.
* We can identify and optimize tech stack diversity.
* We can create awareness and culture shift on certain Legal Entities as needed (e.g., ratio of Open Source and InnerSource below average).

## Rationale

It creates a dynamic and extendable single source of truth for repositories to capture, visualize and act on source code repositories across the Organization. That helps to create awareness and focus efforts on the right direction. The Source Code Strategy Assessment Framework helps teams to understand the value of intentional explicit policies on how to manage source code. It helps to create both continuous improvement cycles and references within the Organization of what others are doing.

## Known Instances

This is under test at scale at:

* Philips

## References

> * Organization and Legal Entity terms as defined in [InnerSource License Pattern Glossary](../2-structured/innersource-license.md#glossary))

Reviewer: ```suggestion
* Organization and Legal Entity terms as defined in [InnerSource License Pattern Glossary](../2-structured/innersource-license.md#glossary).
```

Fixing formatting issue

> - Organization and Legal Entity terms as defined in [InnerSource License Pattern Glossary](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/master/patterns/2-structured/innersource-license.md#glossary))

Reviewer: ```suggestion
- Organization and Legal Entity terms as defined in [InnerSource License Pattern Glossary](../2-structured/innersource-license.md#glossary))
```

Switching to relative links instead.

> - Organization and Legal Entity terms as defined in [InnerSource License Pattern Glossary](../2-structured/innersource-license.md#glossary))
> - Explore aligned deployment of this pattern in combination with the [InnerSource Portal pattern](../2-structured/innersource-portal.md).

Reviewer: ```suggestion
* Organization and Legal Entity terms as defined in [InnerSource License Pattern Glossary](../2-structured/innersource-license.md#glossary))
* Explore aligned deployment of this pattern in combination with the [InnerSource Portal pattern](../2-structured/innersource-portal.md).
```

Using the same bullet point style as in the rest of the doc.

> * Explore aligned deployment of this pattern in combination with the [InnerSource Portal pattern](../2-structured/innersource-portal.md).

Reviewer: ```suggestion
* Explore using this pattern in combination with the [InnerSource Portal](../2-structured/innersource-portal.md) pattern.
```

Does this simplification of this sentence change the intended meaning?

> - Explore aligned deployment of this pattern in combination with the [InnerSource Portal pattern](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/master/patterns/2-structured/innersource-portal.md).

Reviewer: ```suggestion
- Explore aligned deployment of this pattern in combination with the [InnerSource Portal pattern](../2-structured/innersource-portal.md).
```

Switching to relative links instead.

## Title

> Enterprise level source code inventory to drive business value and catalyze InnerSource

Reviewer: I would suggest to use shorter name here, so that it becomes easier to reference the pattern by name from elsewhere. For examples of existing pattern names see [this overview](https://github.com/InnerSourceCommons/InnerSourcePatterns#list-of-patterns).

Ideas for alternative titles:

* Unified source code inventory

## Patlet

> Enterprises sometimes may manage software, in particular source code as key company asset, with less rigorousness than other IT hardware inventories which may diminish the opportunities to maximize business value and control liability cost. An Enterprise level source code inventory can also exploit opportunities to identify and support potential valuable InnerSource assets.

Reviewer: What do you mean by "liability costs" in this context?

Also the Patlet would be stronger if we reduce the conjunctive forms (these forms are sometimes referred to as "timid writers voice").

Here a draft of an alternative version
(not that I already used the terms organization and legal entity as explained below):

> In a large organization with multiple legal entities it is often hard to get full transparency into all software assets, in particular all source code, reducing the opportunities to increase business value and control liability costs across the whole organization. An organization-level source code inventory addresses these issues while also exploiting opportunities to identify and support potentially valuable InnerSource assets.

Author: Liability cost is often times translated into long, increasing and unpredictable maintenance cost for a given software. In other words, if we don't know what we have we can't plan what it will cost to us. I like your proposal. Let me read it in the context and I'll push a new version.

> ## Problem

Reviewer: I would recommend not to use bullet points here but rather prose.

Also if you can tell a story here about a situation in which you encountered this issue, that would be great too. I know it can be tricky for company's to talk about things like this but the more relatable this Problem section is, the easier for others to identify with it.

Author: Good feedback. I've rephrased and added few questions to try to capture interest.

> * Some InnerSource stakeholders do not value source code at the same level of other Enterprise assets.

Reviewer: Any idea why stakeholders don't value source code the same way?

Author: I don't have empirical data or enough sampling. My intuition goes along a general perception around the difficulties of assessing software value for financial capitalization for example. Hardware inventories have an associated value on the balance sheet while software is more intangible.

* Organizations within the Enterprise want to select the right InnerSource projects to maximize outcome.

> * InnerSource Program wants to support the right projects among all the potential candidates.

Reviewer: ```suggestion
* The InnerSource Program wants to support the right projects among all the potential candidates.
```

What qualifies "right projects" here? 
Also why does the InnerSource Program want to do that? i.e. what is their motivation?

* There is no good consolidation of all source code and exploiting business value is ad-hoc.

## Context

* You work on a complex organization under continuous change (e.g., new acquisitions or changing business priorities).
* You cannot find all source code touched, shared and consumed within your organization.
* You do not have a clear policy on default sharing level when creating a new project (e.g., Open, Inner or Closed Source).
* You cannot scan significant parts of your organization's source code looking for duplication, similarity or code smells.
* You do not know the existing ratios of open source, inner source and closed source and their trend.
* You cannot measure the diversity of contributions and resulting value for a given project.
* You cannot identify and optimize tech stack diversity.
* You cannot identify technical debt and determine the priorities for retirement (e.g., dead APIs/source).

> ## Forces

Reviewer: Which of these forces does the source code inventory improve, or improve the most?

Author: It will improve all of them but root causes will not be removed just with this pattern. It will help to create a wrapper around the fragmentation, it will help teams to define strategies and therefore also contributing to change the culture.

Reviewer: Thanks for the extra context 👍

> * Fragmentation of source code hosting systems in your organization.

Reviewer: In [another pattern](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/master/patterns/2-structured/innersource-license.md#glossary) we have introduced these definitions:

**organization** - An umbrella for multiple legal entities. (synonyms: group, enterprise) (e.g. Lufthansa)
**legal entity** - An entity that has its own legal rights and obligations (synonyms: company, subsidiary) (e.g. Lufthansa Systems GmbH, Lufthansa Industry Solutions TS GmbH, ...)

Do you think we could use those in your pattern as well, to stay consistent?

I think we would have to do the following renaming in your pattern:
* organization => legal entity
* enterprise => organization

Author: Great. I'll do. Is there any standard way to cross-refer patters? I could not find a dedicated section in the template but it might be good to avoid duplication on term definition for example.

Author: I've finally added in the references section

Author: I've finally added as a reference.

Reviewer: We don't have a global glossary across all patterns (yet). So what you did is perfect.

As for cross-referencing patterns, what you did works. Sometimes it is also possible to add the references inline somewhere in the pattern itself but either way works.

* Ad-hoc source code strategies scattered across your organization.
* Continuously changing map of the relationships between: projects, repositories, products, tech stacks, domains, solutions, platforms, services, components, sub-systems, people, authors, teams, external repositories.
* Diverse software culture of teams across your organization (e.g., more open to collaboration or more siloed).

## Solutions

> ### Set up an Enterprise level live source code inventory

Reviewer: What do you mean by "live" here?

> * Combination of manual and automated input data sources to a single source of truth including an API for custom extension

Reviewer: What do you mean by "including an API for custom extension"?
Maybe this should be a dedicated bullet in the Solution?

Author: What it means is that together with the backend and frontend implementing the inventory dashboard, there must be a way to add new custom data sources so other teams can merge their inputs. In my experience, fragmentation of hosting can be significant so trying to solve everything from the central team will be challenging. I make it a subbullet for better explanation. Please, review.

* Key meta-data about each repository:
  * Organization
  * URL
  * Version control system (e.g., GIT or SVN).
  * Hosting vendor (e.g., GitHub, Gitlab or BitBucket) and hosting type (e.g., on-prem, private cloud or public cloud).
  * Sharing level (e.g., Open Source, InnerSource, Closed Source)
* Visualization in place to list all assets with options to filter based on meta-data

> * Enable access to automated scanning tools

Reviewer: What do you mean by "scanning tools"?

Author: I mean static analysis to help on code quality. I clarify further with examples.

### Define a source code strategy Framework

* Help to define a source code strategy including an explicit definition of your organization's policy for:
  * Inventory
  * Sharing
  * Consumption
  * Contribution
  * Ecosystem
* Example of such a framework: [https://github.com/trieshard/source-strategy-assessment/blob/master/framework.md](https://github.com/trieshard/source-strategy-assessment/blob/master/framework.md)

## Resulting Context

### For Organizations and Project maintainers

* We have an explicit policy on source code strategy (e.g., where to create new repository or how to select the right sharing level)
* We can find all source code touched, shared and consumed within our Organization and take actions as needed.
* We can measure the diversity of contributions and resulting business value for our project.
* We can identify technical debt and determine the priorities for retirement (e.g., dead APIs/source).

### For InnerSource governance

* We can scan significant parts of our Enterprise's source code looking for opportunities of reuse, duplication, similarity or code smells.

> * We know the ratios of Open, Inner and Closed Source within the Enterprise so we can steer as needed.

Reviewer: ```suggestion
* We know the ratios of Open Source, InnerSource and Closed Source within the Enterprise so we can steer as needed.
```

By spelling out these terms I find it easier to read.

* We can identify and optimize tech stack diversity.

> * We can create awareness and culture shift on certain organizations as needed (e.g., ratio of Open and InnerSource below average)

Reviewer: ```suggestion
* We can create awareness and culture shift on certain organizations as needed (e.g., ratio of Open Source and InnerSource below average)
```

## Rationale

> It creates a dynamic and extendable single source of truth for repositories to capture, visualize and act on source code repositories across your organization. That helps to create awareness and focus efforts on the right direction. The source code maturity framework helps teams to understand the value of intentional explicit policies on how to manage source code. It helps to create both continuous improvement cycles and references within your organization of what others are doing.

Reviewer: Is the "source code maturity framework" here something else than the "source code strategy framework" mentioned above? Or a different name for the same thing?

Also in [the link above](https://github.com/trieshard/source-strategy-assessment/blob/master/framework.md) it is called "Source Code Strategy Assessment Framework", which might be another variant of the same thing? :)

If it is indeed all the same, I would use the exact same name in all the 3 places.

Author: Thanks. Great comment. Yes. Both the same. I've harmonized the use across the document to avoid confusion.
